src/collective/workspace/membership.py

```diff
@@ -90,18 +90,25 @@ def _update_groups(self, old_groups, new_groups, add_auto_groups=True):
         context = workspace.context
         uid = context.UID()
         workspace_groups = get_workspace_groups_plugin()
-
         if self.user is None:
             return
 
         # Determine automatic groups
-        for name, condition in workspace.auto_groups.items():
-            if name not in workspace.available_groups:
-                raise Exception('Unknown workspace group: {}'.format(name))
-            if add_auto_groups and condition(self):
-                new_groups = new_groups.copy() | set([name])
-            else:
-                old_groups = old_groups.copy() | set([name])
+        if add_auto_groups:
+            new_groups = set(new_groups)
+            old_groups = set(old_groups)
+            for name, condition in workspace.auto_groups.items():
+                if name not in workspace.available_groups:
+                    raise Exception('Unknown workspace group: {}'.format(name))
+                if name in new_groups or name in old_groups:
+                    # Respect user input
+                    pass
+                # only add the automatic groups if condition is satisfied,
+                # otherwise remove it
+                if condition(self, new_groups):
```

Comment:
This is a backwards incompatible change in the interface of group conditions. :(

Can you help me understand the scenario where this is needed? `_update_groups` is usually called from `update` *after* it updates self.groups, so the condition should already have access to the new groups.

Reply:
How about we make `_update_groups` start by setting `self.__dict__['groups'] = new_groups` if it's not already equal to that? Then it should always be safe for the condition to use `self.groups`. Of course we should be careful and check that `new_groups` only includes groups which are in `available_groups` but not `auto_groups`.

I know the conditions are new here, but my client has been using them for years now so it would be nice to avoid changing the interface, and I'm not yet convinced that we need to.

Reply:
Uhm, isn't `self.__dict__['groups'] = new_groups` going to remove groups we might want to keep?

To make this patch compatible with the existent code I can change it to:
```
try:
    should_add = condition(self, new_groups)
except TypeError:
    should_add = condition(self)
if should_add:
    ...
```

Reply:
`new_groups` is not a list of groups to add. `new_groups` is the new set of all groups (except for auto groups, until `_update_groups` figures out what they are and adds them).

Maybe it's easier to understand by reading the code I just checked in to the 27-fix-update-groups-davisagli branch...

Reply:
Ah ok! The code you just checked in makes it appear this really clear.

Removing the `add_auto_groups` optional parameter and providing a `_remove_all_groups` method is really an improvement in code readability and maintainability.

I really thought that in the previous implementation `new_groups` were actually the groups to add and `old_groups` the one to remove.
Otherwise I do not see the point in passing the `old_groups` parameter and the lines
https://github.com/collective/collective.workspace/blob/master/src/collective/workspace/membership.py#L107
https://github.com/collective/collective.workspace/blob/master/src/collective/workspace/membership.py#L118

Anyway I see that this  https://github.com/collective/collective.workspace/compare/27-fix-update-groups...27-fix-update-groups-davisagli?expand=1#diff-5f7f9a5a93c00e4d5097f743dd03c7fbR46
indeed fixes the migration issues and I am super happy to have it merged. Do you want to make a new PR for that branch or should I just merge it here?

BTW Thanks for all @davisagli!